File: 0037-sudoku-solver/0037-sudoku-solver.cpp
class Solution {
public:
bool Exist(vector<vector<char>>& board, int r, int c, char val)
{
	for (int i = 0; i < 9; i++)
		if (val == board[r][i])
			return 0;
	for (int i = 0; i < 9; i++)
		if (val == board[i][c])
			return 0;
	for (int i = (r / 3) * 3; i < (r / 3 + 1) * 3; i++)
		for (int j = (c / 3) * 3; j < (c / 3 + 1) * 3; j++)
			if (val == board[i][j])
				return 0;
	return 1;
}
void SudokuSolver(vector<vector<char>>& board, int r, int c, bool& flag)
{
	if (flag)
		return;
	if (board[r][c] != '.')
	{
		++r;
		if (r == 9)
		{
			r = 0;
			++c;
		}
		if (c == 9)
		{
			flag = true;
			return;
		}
		SudokuSolver(board, r, c, flag);
	}
	else
	{
		char val = 49;
		while (!flag && val != 58)
		{
			if (Exist(board, r, c, val))
			{
				board[r][c] = val;
				int i = r + 1;
				int j = c;
				if (i == 9)
				{
					i = 0;
					++j;
				}
				if (j == 9)
				{
					flag = true;
					return;
				}
				SudokuSolver(board, r, c, flag);
				if (!flag)
					board[r][c] = '.';
			}
			++val;
		}
	}
}
    void solveSudoku(vector<vector<char>>& board) {
        bool flag = false;
        SudokuSolver(board,0,0, flag);
        return;
    }
};

File: 0037-sudoku-solver/0037-sudoku-solver.js
const solveSudoku = board => {
    const n = 3, N = n * n;
    const rows = Array.from({ length: N }, () => Array(N + 1).fill(0));
    const cols = Array.from({ length: N }, () => Array(N + 1).fill(0));
    const boxes = Array.from({ length: N }, () => Array(N + 1).fill(0));
    let sudokuSolved = false;

    const couldPlace = (d, row, col) => {
        const idx = Math.floor(row / n) * n + Math.floor(col / n);
        return rows[row][d] + cols[col][d] + boxes[idx][d] === 0;
    };

    const placeNumber = (d, row, col) => {
        const idx = Math.floor(row / n) * n + Math.floor(col / n);
        rows[row][d]++;
        cols[col][d]++;
        boxes[idx][d]++;
        board[row][col] = `${d}`;
    };

    const removeNumber = (d, row, col) => {
        const idx = Math.floor(row / n) * n + Math.floor(col / n);
        rows[row][d]--;
        cols[col][d]--;
        boxes[idx][d]--;
        board[row][col] = '.';
    };

    const placeNextNumbers = (row, col) => {
        if (row === N - 1 && col === N - 1) sudokuSolved = true;
        else if (col === N - 1) backtrack(row + 1, 0);
        else backtrack(row, col + 1);
    };

    const backtrack = (row, col) => {
        if (board[row][col] === '.') {
            for (let d = 1; d <= 9; d++) {
                if (couldPlace(d, row, col)) {
                    placeNumber(d, row, col);
                    placeNextNumbers(row, col);
                    if (!sudokuSolved) removeNumber(d, row, col);
                }
            }
        } else placeNextNumbers(row, col);
    };

    for (let i = 0; i < N; i++)
        for (let j = 0; j < N; j++)
            if (board[i][j] !== '.') placeNumber(Number(board[i][j]), i, j);

    backtrack(0, 0);
};